Add uvisor support to Beetle

fvincenzo opened the pull request:
This PR contains two patches and adds the support for Beetle to uvisor.

meriac commented on core/debug/src/debug.c:
```diff
@@ -30,6 +30,41 @@ uint8_t g_buffer[DEBUG_MAX_BUFFER];
 int g_buffer_pos;
 
 #ifndef CHANNEL_DEBUG
```
@fvincenzo please move all the beetle-specific code to a dedicated debug_beetle.c file.

fvincenzo replied:
Done :smile:

meriac commented on core/debug/src/debug.c:
```diff
@@ -30,6 +30,41 @@ uint8_t g_buffer[DEBUG_MAX_BUFFER];
+
+    char c = (char) data;
+
+    while(beetle_uart1_writable());
```
@fvincenzo name looks wrong: I assume you spin while *not* writable.

meriac commented on core/debug/src/debug.c:
```diff
@@ -30,6 +30,41 @@ uint8_t g_buffer[DEBUG_MAX_BUFFER];
+#define UART1BDIV   (UART1 + 0x10)
+
+#define beetle_uart1_readl(reg) *(volatile uint32_t *)reg
+#define beetle_uart1_writel(reg, val)  *(uint32_t *)reg = val;
```
@fvincenzo no good reason to use macros here: can you please use static inlines instead?

fvincenzo replied:
ok

meriac commented:
@AlessandroA please review, too.

meriac commented on core/debug/src/debug_beetle.c:
```diff
@@ -0,0 +1,67 @@
+
+static inline void beetle_uart1_writel(uint32_t reg, uint32_t val)
+{
+    *(uint32_t *)reg = val;
```
@fvincenzo "volatile" missing here?

meriac replied:
@fvincenzo form the looks of it you would be better off simply declaring a struct and casting it to the absolute offset. Each struct member would be declared with _IO _R _W as seen in the CMSIS headers - which would solve the "volatile"- problems.

meriac replied:
@fvincenzo (a UART struct, casted to be at UART1 position)

AlessandroA commented on platform/beetle/inc/config.h:
```diff
@@ -0,0 +1,27 @@
+#define __CONFIG_H__
+
+/* uVisor stack configuration */
+#define STACK_SIZE       2048
```
@fvincenzo These 2 lines could be removed. Could you please verify that it builds fine without them? (They should all have a default value).

AlessandroA commented:
@fvincenzo For the `debug_beetle.c` file what I suggest to do is to put it into `platform/beetle/src/debug_beetle.c`. If I remember correctly the build system si already able to detect platform-specific code (let me know if it doesn't).

Once the code is in the Beetle-specific folder, you remove the use of the `CORE_BEETLE` symbol from the whole code-base (in `debug.c`, in the configuration file, and in the `debug_beetle.c` file).

fvincenzo commented:
@AlessandroA 
I modified the PR based on your suggestions and rebased on the latest main. Unfortunately if I build in this condition uvisor stops working.
```
uVisor mode: 2
uvisor_ram : @0x20000140 (7872 bytes) [config]
             @0x20000140 (7872 bytes) [linker]
bss_boxes  : @0x20002000 (8320 bytes) [config]

box stack segment start=0x20002000 end=0x20002000 (length=8320)
MPU.REGIONS=8
MPU.ALIGNMENT=0x0000001F
MPU.ALIGNMENT_BITS=5
HALT_ERROR(./core/system/src/mpu/vmpu_armv7m_mpu.c#178): inferred ACL's (0x003F) don't match exptected ACL's (0x0037)
```

This issue seems not depending from my code.

AlessandroA commented:
@fvincenzo I can confirm this issue was already observed by @niklas-arm in his port to the EFM32GG_STK3700 module. I'll raise an issue here and verify the ACLs independently from this PR. Thanks for the update!
